### PR TITLE
HDDS-2233 - Remove ByteStringHelper and refactor the code to the place where it used

### PR DESCRIPTION
**Pull request by fapifta**
There are a couple of things in this pull request.
ByteStringHelper was used in two general places, once in BlockOutputStream, and once in ChunkUtils.
ChunkUtils was an easy one, as in that case the method was used directly from KeyValueHandler only, where the configuration is present. I moved the logic back to the KeyValueHandler class, and at initialization time, I created the proper converter based on the config.
At this point I realized that we use byte[] type in readChunk like methods here and there, but at the end most of the usage outside is via a ByteBuffer after a ByteBuffer.wrap(byte[]) call, so I changed the readChunk logic to return a ByteBuffer instead, and used that one, with this I eliminated the need of a conversion from byte[] to ByteString.

BlockOutputStream was a bit harder, as in that case, RPCClient initialized the old class with the config. The decision became the following: based on the Configuration, RPCClient already created an XCeiverClientManager that preserved a reference to the configuration, and which was already available in the BlockOutputStreamEntryPool class, that uses a BufferPool to create the buffers in BlockOutputStream, so that BlockOutputStream when does the conversion can reach it via its BufferPool.

I have cleaned up one code path, that came into my way during implementation, and I removed two declared exception from ChunkUtils.readData, and with that eliminated the need for a try-catch block in the place where it is called.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 65 | Maven dependency ordering for branch |
| -1 | mvninstall | 40 | hadoop-hdds in trunk failed. |
| -1 | mvninstall | 39 | hadoop-ozone in trunk failed. |
| -1 | compile | 21 | hadoop-hdds in trunk failed. |
| -1 | compile | 15 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 59 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 854 | branch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 20 | hadoop-hdds in trunk failed. |
| -1 | javadoc | 18 | hadoop-ozone in trunk failed. |
| 0 | spotbugs | 951 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 32 | hadoop-hdds in trunk failed. |
| -1 | findbugs | 22 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 30 | Maven dependency ordering for patch |
| -1 | mvninstall | 35 | hadoop-hdds in the patch failed. |
| -1 | mvninstall | 37 | hadoop-ozone in the patch failed. |
| -1 | compile | 23 | hadoop-hdds in the patch failed. |
| -1 | compile | 17 | hadoop-ozone in the patch failed. |
| -1 | javac | 23 | hadoop-hdds in the patch failed. |
| -1 | javac | 17 | hadoop-ozone in the patch failed. |
| +1 | checkstyle | 55 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 701 | patch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 21 | hadoop-hdds in the patch failed. |
| -1 | javadoc | 19 | hadoop-ozone in the patch failed. |
| -1 | findbugs | 33 | hadoop-hdds in the patch failed. |
| -1 | findbugs | 20 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 27 | hadoop-hdds in the patch failed. |
| -1 | unit | 25 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 2437 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1596 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 954300498c97 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 6171a41 |
| Default Java | 1.8.0_222 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-compile-hadoop-hdds.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/testReport/ |
| Max. process+thread count | 446 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/client hadoop-hdds/common hadoop-hdds/container-service hadoop-ozone/client hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1596/1/console |
| versions | git=2.7.4 maven=3.3.9 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by fapifta** [comment]
Acceptance test failures seems to be related to an internally examined similar failure in XCeiverClientGrpc, where the relevant stack trace is:
Caused by: java.lang.NullPointerException
    at org.apache.hadoop.hdds.scm.XceiverClientGrpc.sendCommandAsync(XceiverClientGrpc.java:387)
    at org.apache.hadoop.hdds.scm.XceiverClientGrpc.sendCommandWithRetry(XceiverClientGrpc.java:285)
    at org.apache.hadoop.hdds.scm.XceiverClientGrpc.sendCommandWithTraceIDAndRetry(XceiverClientGrpc.java:251)
    at org.apache.hadoop.hdds.scm.XceiverClientGrpc.sendCommand(XceiverClientGrpc.java:234)
    at org.apache.hadoop.hdds.scm.storage.ContainerProtocolCalls.readChunk(ContainerProtocolCalls.java:245)
    at org.apache.hadoop.hdds.scm.storage.ChunkInputStream.readChunk(ChunkInputStream.java:335)
    at org.apache.hadoop.hdds.scm.storage.ChunkInputStream.readChunkFromContainer(ChunkInputStream.java:307)
    at org.apache.hadoop.hdds.scm.storage.ChunkInputStream.prepareRead(ChunkInputStream.java:249)
    at org.apache.hadoop.hdds.scm.storage.ChunkInputStream.read(ChunkInputStream.java:144)
    at org.apache.hadoop.hdds.scm.storage.BlockInputStream.read(BlockInputStream.java:239)
    at org.apache.hadoop.ozone.client.io.KeyInputStream.read(KeyInputStream.java:171)
    at org.apache.hadoop.fs.ozone.OzoneFSInputStream.read(OzoneFSInputStream.java:52)
    at org.apache.hadoop.fs.FSInputStream.read(FSInputStream.java:75)
    at org.apache.hadoop.fs.FSInputStream.readFully(FSInputStream.java:121)
    at org.apache.hadoop.fs.FSDataInputStream.readFully(FSDataInputStream.java:112)
    at org.apache.orc.impl.ReaderImpl.extractFileTail(ReaderImpl.java:555)

From the Integration test failures org.apache.hadoop.ozone.client.rpc.TestFailureHandlingByClient failure might be related, but there in the output before the error we can see that the client wanted replication factor of 3, while the cluster had only 1 DN. 

So issues seem to be unrelated, however I can not run these tests locally as easy as they are timing out in my local machine.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/ByteStringConversion.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    boolean unsafeEnabled =
+        config!=null && config.getBoolean(
+            OzoneConfigKeys.OZONE_UNSAFEBYTEOPERATIONS_ENABLED,
+            OzoneConfigKeys.OZONE_UNSAFEBYTEOPERATIONS_ENABLED_DEFAULT);
+    if (unsafeEnabled) {
+      return buffer -> UnsafeByteOperations.unsafeWrap(buffer);
+    } else {
+      return buffer -> {
+        ByteString retval = ByteString.copyFrom(buffer);
+        buffer.flip();
```
Comment (mukul1987) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/ByteStringConversion.java]: Why is the buffer.flip needed ? In the earlier ByteString helper class this change wasn't needed.

Comment (bshashikant) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/ByteStringConversion.java]: Buffer.flip() will be required to reset the pos to 0, as ByteString.Copy advances the position to the data length,

**Comment by bshashikant** [comment]
Thanks @fapifta for working on this. In addition to Mukul's comments, I feel ByteBuffer to byteString conversion should be a general utility function and should not take configuration object as dependency to determine safe or unsafe copy. Let's pass the boolean flag instead as it would be more friendly to use where ever needed.

**Comment by fapifta** [comment]
Hi @bshashikant 

first of all, thank you for the review.

Let me share my thought process when I selected to have a utility method that is dependent on the Configuration we have:
- the logic of conversion is changing based on a configuration value
- we don't use the utility method so far in any other context where the decision of which implementation is chosen is made based on anything else then configuration
- this way I need to write the reading and handling of the related Configuration value once
- this coupling of the implementation choosing logic and configuration I think is best to be encapsulated into the utility method that creates our conversion strategy, to express it clearly that this is dependent on a configuration value in the current environment.


So all in all I would refuse to have the method with a boolean parameter for now as I don't see its benefits at the moment, and also because if there will ever be an other arbitrary decision making process coexisting besides the using of the Configuration, then I think we can factor a method out from the current one, which will accept a boolean as a parameter, or based on other parameters we can implement the decision logic for other type of clients when there is a need.
Does this sound reasonable and acceptable?

**Comment by fapifta** [comment]
@mukul1987 I have created https://issues.apache.org/jira/browse/HDDS-2261 to describe the byte[]->ByteBufffer conversion, and linked this pull request there as well, let me know if I missed something about it. Thank you for the suggestion.

**Comment by anuengineer** [comment]
+1. LGTM.  I will wait for others to finish commenting to commit this.

**Comment by bshashikant** [comment]
Thanks @fapifta for the clarification. I am +1 on the change as well.

**Comment by bshashikant** [comment]
Thanks @fapifta for workin on this. I have committed this.

**Pull request merged**